### DISCUSSION TITLE
[@types/node] ReadableStream global should be an AsyncIterable

### DISCUSSION THREAD

**Discussion by nickserv** [DefinitelyTyped/DefinitelyTyped#62651] [Issues with a @types package]
The Node `ReadableStream` implementation implements the `AsyncIterable` interface:
```js
ReadableStream.prototype[Symbol.asyncIterator]
// [Function: values]
```
However, using `ReadableStream` globally (without importing from `stream/web`) results in a type error that doesn't adhere to the spec:
```ts
for await (const value of new ReadableStream()) {}
// error TS2504: Type 'ReadableStream<any>' must have a '[Symbol.asyncIterator]()' method that returns an async iterator.
```

[a comment by an automated account is not shown]

**Comment by MysteryBlokHed** [comment]
The global version of ReadableStream is the one from the DOM types, which doesn't define `Symbol.asyncIterator`. Importing it from `node:stream/web` or using the type `NodeJS.ReadableStream` should work

**Reply by nickserv** [reply]
Yes, but this isn't consistent with Node supporting `Symbol.asyncIterator` on the global, and we can't add this to TypeScript core until it's been implemented in two browsers. So I think we should just patch the global types here, considering that users already expect globals like `require`, `module`, and `process` to be patched in.

**Reply by brandonryan** [reply]
Is this still true? On the MDN documentation for Web Streams, it documents the [for await usage](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#async_iteration). Can this be added now?

**Comment by Holi0317** [comment]
I am facing this problem where some api (`Readable.fromWeb`) should work in runtime but typescript think it is not.

In the end, I am solving this by adding a shim/override in my project's `global.d.ts` to convince typescript the global `ReadableStream` defined in `lib.dom.d.ts` is compatible with node's implementation.

```ts
interface ReadableStream<R = any> {
  values(options?: { preventCancel?: boolean }): AsyncIterableIterator<R>;
  [Symbol.asyncIterator](): AsyncIterableIterator<R>;
}

```

**Comment by beorn** [comment]
Would be nice to bump this up in priority for Chrome and WebKit - with streams gaining popularity and being standard, it is a shame you can't use idiomatic `for await` looping over streams:
- https://bugs.chromium.org/p/chromium/issues/detail?id=929585
- https://bugs.webkit.org/show_bug.cgi?id=194379
